File: solutions/clojure/resistor-color-trio/1/src/resistor_color_trio.clj
(ns resistor-color-trio)

(def color-map {"black" 0, "brown" 1, "red" 2, "orange" 3, 
                "yellow" 4, "green" 5, "blue" 6, "violet" 7,
                "grey" 8, "white" 9})
(def names-map {0 "ohms", 1 "kiloohms", 2 "megaohms", 3 "gigaohms"})
(def scales-map {0 1, 1 1000, 2 1000000, 3 1000000000})

(defn get-scale [ohms]
  (cond 
      (>= ohms (get scales-map 3)) 3
      (>= ohms (get scales-map 2)) 2
      (>= ohms (get scales-map 1)) 1
      :else 0
    ))

(defn scaled-value [ohms scale]
    (/ ohms (get scales-map scale)))

(defn get-value [colors]
  (* (+ (* 10 (get color-map (nth colors 0))) 
              (get color-map (nth colors 1)))
     (int (Math/pow  10 (get color-map (nth colors 2))))))

(defn resistor-label
  "Returns the resistor label based on the given color bands."
  [colors]
    (let [ohms (get-value colors)
          scale (get-scale ohms)]
      (clojure.string/join " " [(str (scaled-value ohms scale)) 
                   (get names-map scale)])))
  

File: solutions/java/resistor-color-trio/1/src/main/java/ResistorColorTrio.java
import java.util.Map;

class ResistorColorTrio {
    static private Map<String, Integer> ccodeMap = Map.ofEntries(
        Map.entry("black", 0),
        Map.entry("brown", 1),
        Map.entry("red", 2),
        Map.entry("orange",3),
        Map.entry("yellow",4),
        Map.entry("green", 5),
        Map.entry("blue",6),
        Map.entry("violet",7),
        Map.entry("grey",8),
        Map.entry("white",9)
   );
    
    String label(String[] colors) {
        int n1 = ccodeMap.get(colors[0]);
        int n2 = ccodeMap.get(colors[1]);
        int n3 = ccodeMap.get(colors[2]);
        int ohms = 10*n1 + n2;
        String color_label = "";
        if ((n3<=1) || (n1 ==0 && n3==2)){
            ohms *= n3 == 2 ? 100 : n3 == 1 ? 10 : 1;
            color_label = Integer.toString(ohms)+" ohms";
        }
        else if (n1!=0 && n2==0 && n3==2){
            color_label = Integer.toString(n1)+" kiloohms";
        }
        else if (n1!=0 && n2==0 && n3==2){
            color_label = Integer.toString(n1)+" kiloohms";
        }
        else if (n1!=0 && n2!=0 && n3==2){
            ohms /= 10;
            color_label = Integer.toString(ohms)+" kiloohms";
        }
        else if (n3<=4 || (n1 == 0 && n3 == 5)){
            ohms *= n3==5 ? 100 : n3==4 ? 10 : 1; 
            color_label = Integer.toString(ohms)+" kiloohms";
        }
        else if (n1 !=0 && n2 ==0 && n3==5){
            color_label = Integer.toString(n1)+" megaohms";
        }
        else if (n3<=7 || (n1 == 0 && n3 == 8)){
            ohms *= n3==8 ? 100 : n3==7 ? 10 : 1; 
            color_label = Integer.toString(ohms)+" megaohms";
        }
        else if (n1!=0 && n2 !=0 && n3== 8){
            ohms /= 10;
            color_label = Integer.toString(ohms)+" gigaohms";
        }
        else {
            color_label = Integer.toString(ohms)+" gigaohms";
        }
        return color_label;
    }
}
